paddle/gserver/tests/ProtoDataProvider2.cpp

```diff
@@ -0,0 +1,557 @@
+      if (memUsage > FLAGS_memory_threshold_on_load_data) {
+        LOG(INFO) << "memUsage is " << memUsage << ", > "
+        << FLAGS_memory_threshold_on_load_data
+        << " therefore SKIP ALL REMAINING file.";
+        break;
+      }
+    }
+    LOG(INFO) << "load data file " << file;
+    loadDataFile(file);
+  }
+  LOG(INFO) << "read done, num of instance=" << sampleNums_;
+}
+
+void ProtoDataProvider2::loadData(const std::string& fileName) {
+  std::vector<std::string> fileList;
+  loadFileList(fileName, fileList);
+  loadData(fileList);
+}
+
+
+void ProtoDataProvider2::initDataHeader(const DataHeader2& header) {
+  std::call_once(once_flag_, [&] {
+    header_ = header;
+    this->scanners_.resize(header_.slot_defs_size());
+    for (int i = 0; i < header_.slot_defs_size(); ++i) {
+      this->scanners_[i].reset(
+              ISlotScanner::create(header_.slot_defs(i), header_.seq_type(i)));
+    }
+  });
+}
+
+void ProtoDataProvider2::loadDataFile(const std::string& fileName) {
+  std::ifstream is(fileName);
+  CHECK(is) << "Fail to open " << fileName;
+  bool dataCompression = str::endsWith(fileName, ".gz");
+  std::unique_ptr<ProtoReader> reader(new ProtoReader(&is, dataCompression));
+  CHECK(reader) << "Fail to create proto data input stream";
+
+  DataHeader2 header;
+  CHECK(reader->read(&header));
+  initDataHeader(header);
+
+
+  do {
+    DataSample2* samplePtr = new DataSample2();
+    DataSample2& sample = *samplePtr;
+    if (!reader->read(&sample)) {
+      break;
+    }
+    sample_.push_back(std::shared_ptr<DataSample2>(samplePtr));
+    ++sampleNums_;
+  } while (true);
+
+  CHECK(is.eof()) << "Fail to read file";
+  reader.reset(nullptr);
+  is.close();
+}
+
+
+int64_t ProtoDataProvider2::getNextBatchInternal(int64_t size,
+                                                 DataBatch *batchOut) {
+  int64_t actualSize = 0;
+  std::unique_lock<std::mutex> guard(lock_);
+  std::vector<std::shared_ptr<DataSample2>> buf;
+  buf.reserve(size);
+  for (; actualSize < size &&
+         currentIndex_ + actualSize < static_cast<int64_t>(sampleNums_);
+          ++actualSize) {
+    buf.push_back(sample_[currentIndex_ + actualSize]);
+  }
+  currentIndex_ += actualSize;
+  if (actualSize > 0) {
+    DataBatch& cpuBatch = *cpuBatch_;
+    process(buf, &cpuBatch);
+    if (useGpu_) {
+      std::vector<Argument>& cpuArgs = cpuBatch.getStreams();
+      DataBatch& gpuBatch = *gpuBatch_;
+      std::vector<Argument>& gpuArgs = gpuBatch.getStreams();
+      gpuArgs.resize(cpuArgs.size());
+      gpuBatch.setSize(size);
+      for (size_t i = 0; i < cpuArgs.size(); ++i) {
+        gpuArgs[i].resizeAndCopyFrom(cpuArgs[i], useGpu_, HPPL_STREAM_1);
+      }
+      hl_stream_synchronize(HPPL_STREAM_1);
+       *batchOut = gpuBatch;
+    } else {
+      *batchOut = cpuBatch;
+    }
+  }
+  return actualSize;
+}
+
+void ProtoDataProvider2::shuffle() {
+  std::random_shuffle(sample_.begin(), sample_.end());
```

Comment:
It might be a tiny thing, but `std::random_shuffle` is deprecated in `c++ 14`.
See http://stackoverflow.com/questions/22600100/why-are-stdshuffle-methods-being-deprecated-in-c14 .

Use `std::shuffle(sample_.begin(), sample_.end(), ThreadLocalRandomEngine::get())` instead.